README
HTML::Lint is a pure-Perl HTML parser and checker for syntactic legitmacy.

The "weblint" script is now pretty much a wrapper around the HTML::Lint
module.

You can also look into Apache::Lint which is a mod_perl wrapper around
HTML::Lint.

Finally, for those of you doing automated testing with Test::More and
its brethren, Test::HTML::Lint lets you automate HTML checking.

Please let me know if any tests fail.

Andy Lester
andy at petdance.com


Changes
Revision history HTML::Lint and Test::HTML::Lint.

NOTE: All bugs and requests are now being handled through GitHub.

    https://github.com/petdance/html-lint/issues

Please DO NOT send bug reports to http://rt.cpan.org/.


2.32    Fri Jun 22 15:57:39 CDT 2018
    [ENHANCEMENTS]
    Allow "weblint -" to read from STDIN.  Thanks, Frank Dana.


2.30    Sun Jan  7 22:02:25 CST 2018
    No changes since 2.27_03.


2.27_03 Wed Jan  3 17:07:07 CST 2018
    [FIXES]
    Removed the text-invalid-entity and attr-invalid-entity, which were
    for entities that had an invalid numeric value, anything greater
    than &#65536.  There is no longer a restriction on the numeric values
    of HTML entities. (GH#60)


2.27_02 Wed Dec 27 11:46:28 CST 2017
    There are be no functionality changes since 2.27_01.

    [INTERNALS]
    Many Perl::Critic cleanups.


2.27_01 Fri Dec 22 15:54:32 CST 2017
    [ENHANCEMENTS]
    Adds checking of entities in attributes, not just text.  Thanks,
    Klaus S. Madsen.

    [FIXES]
    Calling ->parsefile() would generate an error.  Thanks, Shlomi
    Fish. (GH#58)

    [INTERNALS]
    Prepare for perl 5.26.0 which removes '.' from @INC.  Thanks,
    Jim Keenan.

    Fix disttest target.  Thanks, Shlomi Fish.


2.26    Thu Dec 29 22:36:54 CST 2016
    Stable release.  No changes from previous release.


2.25_02 Tue Dec 27 14:34:22 CST 2016
    [FIXES]
    html_fragment_ok() was not properly excluding document-level errors.
    It was effectively the same as html_ok().


2.25_01 Fri Dec 23 22:36:17 CST 2016
    [ENHANCEMENTS]
    Added two new types of errors to let you know you're using the
    API incorrectly.  You should be parsing files like this:

        my $lint = HTML::Lint->new;
        $lint->newfile( $filename );
        $lint->parse( $line );
        $lint->eof();
        my @errors = $lint->errors();

    If you neglect to call ->parse or ->eof, you'll get an error returned
    in the list of errors from ->errors().

    [FIXES]
    Test::HTML::Lint::html_fragment_ok() was not properly calling ->eof.


2.24    Wed Dec  7 22:20:13 CST 2016
    Official release.  No changes from 2.23_01.


2.23_01 Tue Dec  6 22:48:56 CST 2016
    [ENHANCEMENTS]
    Added detection of unknown HTML entities, like "known &unclosed
    &entities are not found".  Also fixes the case where HTML::Lint
    gets confused by an entity like "&sup2;" which it thought was an
    unterminated "&sup;" entity.  Thanks, Klaus S. Madsen.

    [FIXES]
    Errors of the type doc-tag-required did not come out in any defined
    order.  They are now sorted by tag name.  This was discovered
    because hash randomization caused tests to fail on Perl 5.18 and
    above. Thanks, Slaven Rezic, Andrew Main and Lisa Hare.

    Handle some warnings that get thrown if certain values are undef.
    Thanks, Yves Lavoie.

    Handle characters that are not handled by HTML::Entities. (GitHub
    issue #13) Thanks, Tim Landscheidt.

    [INTERNALS]
    Add a test to verify a fixed bug. Thanks to Lance Wicks as part of
    the CPAN Pull Request Challenge.


2.22    Mon Apr  6 15:47:11 CDT 2015
    [CHANGES THAT COULD BREAK YOUR CODE]
    Previously, html_ok() would not check the entire structure of a web
    page to check for <html>, <head>, <title> and <body> tags.  Now it
    will.  If you want to check fragments of HTML for validity but know
    that they are not valid HTML documents on their own, use the new
    html_fragment_ok().

    [ENHANCEMENTS]
    Added new error, elem-input-alt-missing, that warns of <input
    type="image"> tags that are missing an alt="" attribute.  This helps
    for accessability to make sure that any images have alternate text
    for screen readers.

    Added ability to modify HTML::Lint's table of known tags and
    attributes, so you could do this:

        # Add an attribute that your company uses.
        HTML::Lint::HTML4::add_attribute( 'body', 'proprietary-attribute' );

        # Add the HTML 5 <canvas> tag.
        HTML::Lint::HTML4::add_tag( 'canvas' );
        HTML::Lint::HTML4::add_attribute( 'canvas', $_ ) for qw( height width );

    [FIXES]
    Test::HTML::Lint::html_ok() would not call the HTML::Lint eof()
    method, which meant it wouldn't do document-wide tests.

2.20    Fri Apr  6 00:49:51 CDT 2012
    [ENHANCEMENTS]
    Sometimes creating HTML::Lint-compliant HTML just isn't possible.
    Now, you can now turn individual errors on and off in your HTML
    via comment directives, like so:

        <!-- html-lint elem-img-sizes-missing: off, attr-unknown: off -->

    And if you have a batch of code that's hopeless:

        <!-- html-lint all: off -->

    Added check for unknown entities, such as "&foo;".

    Added check for unclosed entitities, such as "&amp" without the
    closing semicolon.

    Added a check for a bare ampersand that should be written as &amp;


2.10    Tue Dec  6 11:16:16 CST 2011
    [FIXES]
    Tags that were self-closed were being ignored.  For example, if you had

        <img src="blah.jpg" />

    then HTML::Lint would ignore the tag.  This has been fixed.

    [LICENSE]
    HTML::Lint is now explicitly licensed under Artistic License 2.0,
    instead of the vague "same terms as Perl itself."


2.06    Thu Dec 18 00:07:54 CST 2008
    [FIXES]
    Added attributes to <frameset>.

    <strong> tag didn't allow any attributes.  Now it does.

    Removed the <listing> tag.


2.04    Mon Jun  2 11:41:16 CDT 2008
    [FIXES]
    <textarea> now knows about the wrap attribute.

    [INTERNALS]
    Lots of enhancements pointed out by Perl::Critic.


2.02    Thu Nov  3 11:49:18 CST 2005
    [ENHANCEMENTS]
    * The warnings for missing ALT and HEIGHT/WIDTH on your images
        now give the SRC attribute.


2.00    Tue Sep 20 23:10:39 CDT 2005
    [CHANGES THAT COULD BREAK YOUR CODE]
    * I've changed the object structure.  HTML::Lint now has-a
        HTML::Parser, and no longer is-a HTML::Parser.


t/Util.pm
package Util;

use parent 'Exporter';

use warnings;
use strict;

use Test::More;
use HTML::Lint;

our @EXPORT = qw(
    checkit
);

sub checkit {
    my @expected = @{+shift};
    my @linesets = @_;

    plan( tests => 3*(scalar @expected) + 4 );

    my $lint = new HTML::Lint;
    isa_ok( $lint, 'HTML::Lint', 'Created lint object' );

    my $n;
    for my $set ( @linesets ) {
        ++$n;
        $lint->newfile( "Set #$n" );
        $lint->parse( $_ ) for @{$set};
        $lint->eof;
    }

    my @errors = $lint->errors();
    is( scalar @errors, scalar @expected, 'Right # of errors' );

    while ( @errors && @expected ) {
        my $error = shift @errors;
        isa_ok( $error, 'HTML::Lint::Error' );

        my $expected = shift @expected;

        is( $error->errcode, $expected->[0], 'Error codes match' );
        my $match = $expected->[1];
        if ( ref($match) eq 'Regexp' ) {
            like( $error->as_string, $match, 'Error matches regex' );
        }
        else {
            is( $error->as_string, $match, 'Error matches string' );
        }
    }

    my $dump;

    is( scalar @errors, 0, 'No unexpected errors found' ) or $dump = 1;
    is( scalar @expected, 0, 'No expected errors missing' ) or $dump = 1;

    if ( $dump && @errors ) {
        diag( 'Leftover errors...' );
        diag( $_->as_string ) for @errors;
    }
}

1; # happy


t/LintTest.pl
use Test::More;
use HTML::Lint;

sub checkit {
    my @expected = @{+shift};
    my @linesets = @_;

    plan( tests => 2*(scalar @expected) + 4 );

    my $lint = new HTML::Lint;
    isa_ok( $lint, 'HTML::Lint', 'Created lint object' );

    my $n;
    for my $set ( @linesets ) {
        ++$n;
        $lint->newfile( "Set #$n" );
        $lint->parse( $_ ) for @$set;
        $lint->eof;
    }

    my @errors = $lint->errors();
    is( scalar @errors, scalar @expected, 'Right # of errors' );

    my $expected_found = 0;
    while ( @errors ) {
        my $error = shift @errors;
        isa_ok( $error, 'HTML::Lint::Error' );
        foreach my $exp (@expected) {
            my $match = $exp->[1];
            if ( ref $match eq "Regexp" && $error->as_string =~ /$match/) {
                $expected_found++;
                is( $error->errcode, $exp->[0], 'Error codes match' );
                last;
            }
            elsif ($error->as_string eq $match) {
                $expected_found++;
                is( $error->errcode, $exp->[0], 'Error codes match' );
                last;
            }

        }
    }

    my $dump;
    is( scalar @errors, 0, 'No unexpected errors found' ) or $dump = 1;
    is( $expected_found, scalar @expected, 'No expected errors missing' ) or $dump = 1;

    if ( $dump && @errors ) {
        diag( "Leftover errors..." );
        diag( $_->as_string ) for @errors;
    }
}

# Read in a set of sets of lines, where each "file" is separated by a
# blank line in <DATA>
sub get_paragraphed_files {
    local $/ = "";

    my @sets;

    while ( my $paragraph = <DATA> ) {
        my @lines = split /\n/, $paragraph;
        @lines = map { "$_\n" } @lines;
        push( @sets, [@lines] );
    }

    return @sets;
}

1; # happy


lib/Test/HTML/Lint.pm
package Test::HTML::Lint;

use warnings;
use strict;

use Test::Builder;
use Exporter;

use HTML::Lint;

use vars qw( @ISA $VERSION @EXPORT );

@ISA = qw( HTML::Parser Exporter );

=head1 NAME

Test::HTML::Lint - Test::More-style wrapper around HTML::Lint

=head1 VERSION

Version 2.28

=cut

$VERSION = '2.28';

my $Tester = Test::Builder->new;

=head1 SYNOPSIS

    use Test::HTML::Lint tests => 4;

    my $table = build_display_table();
    html_ok( $table, 'Built display table properly' );

=head1 DESCRIPTION

This module provides a few convenience methods for testing exception
based code. It is built with L<Test::Builder> and plays happily with
L<Test::More> and friends.

If you are not already familiar with L<Test::More> now would be the time
to go take a look.

=head1 EXPORT

C<html_ok>

=cut

@EXPORT = qw(
    html_ok
    html_fragment_ok
);

sub import {
    my $self = shift;
    my $pack = caller;

    $Tester->exported_to($pack);
    $Tester->plan(@_);

    $self->export_to_level(1, $self, @EXPORT);

    return;
}

=head2 html_ok( [$lint, ] $html, $name )

Checks to see if C<$html> is a valid HTML document, including checks
for having C<< <html> >>, C<< <head> >>, C<< <title>> >> and
C<< <body> >> tags.

If you're checking something that is only a fragment of an HTML document,
use C<html_fragment_ok()>.

If you pass an HTML::Lint object, C<html_ok()> will use that for its
settings.

    my $lint = new HTML::Lint( only_types => STRUCTURE );
    html_ok( $lint, $content, "Web page passes structural tests only" );

Otherwise, it will use the default rules.

    html_ok( $content, "Web page passes ALL tests" );

Note that if you pass in your own HTML::Lint object, C<html_ok()>
will clear its errors before using it.

=cut

sub html_ok {
    my $lint;

    if ( ref($_[0]) eq 'HTML::Lint' ) {
        $lint = shift;
        $lint->newfile();
        $lint->clear_errors();
    }
    else {
        $lint = HTML::Lint->new;
    }
    my $html = shift;
    my $name = shift;

    my $ok = defined $html;
    if ( !$ok ) {
        $Tester->ok( 0, $name );
    }
    else {
        $lint->parse( $html );
        $lint->eof();
        my $nerr = scalar $lint->errors;
        $ok = !$nerr;
        $Tester->ok( $ok, $name );
        if ( !$ok ) {
            my $msg = 'Errors:';
            $msg .= " $name" if $name;
            $Tester->diag( $msg );
            $Tester->diag( $_->as_string ) for $lint->errors;
        }
    }

    return $ok;
}


=head2 html_fragment_ok( [$lint, ] $html, $name )

Checks that C<$fragment> is valid HTML, but not necessarily a valid
HTML document.

For example, this is a valid fragment, but not a valid HTML document:

    <body>
        <p>Lorem ipsum</p>
    </body>

because it doesn't contain C<< <html> >> and C<< <head> >> tags.  If you
want to check that it is a valid document, use C<html_ok()>.

If you pass an HTML::Lint object, C<html_fragment_ok()> will use that for its
settings.

    my $lint = new HTML::Lint( only_types => STRUCTURE );
    html_fragment_ok( $lint, $content, 'Web page passes structural tests only' );

Otherwise, it will use the default rules.

    html_fragment_ok( $content, 'Fragment passes ALL tests' );

Note that if you pass in your own HTML::Lint object, C<html_fragment_ok()>
will clear its errors before using it.

=cut

sub html_fragment_ok {
    my $lint;

    if ( ref($_[0]) eq 'HTML::Lint' ) {
        $lint = shift;
        $lint->newfile();
        $lint->clear_errors();
    }
    else {
        $lint = HTML::Lint->new;
    }
    my $html = shift;
    my $name = shift;

    my $ok = defined $html;
    if ( !$ok ) {
        $Tester->ok( 0, $name );
    }
    else {
        $lint->parse( $html );
        $lint->eof();

        # Ignore doc-level errors.
        my @errors = grep { $_->errcode ne 'doc-tag-required' } $lint->errors;

        my $nerr = @errors;
        $ok = !$nerr;
        $Tester->ok( $ok, $name );
        if ( !$ok ) {
            my $msg = 'Errors:';
            $msg .= " $name" if $name;
            $Tester->diag( $msg );
            $Tester->diag( $_->as_string ) for @errors;
        }
    }

    return $ok;
}


=head1 BUGS

All bugs and requests are now being handled through GitHub.

    https://github.com/petdance/html-lint/issues

DO NOT send bug reports to http://rt.cpan.org/.


=head1 TO DO

There needs to be a C<html_table_ok()> to check that the HTML is a
self-contained, well-formed table, and then a comparable one for
C<html_page_ok()>.

If you think this module should do something that it doesn't do at the
moment please let me know.

=head1 ACKNOWLEDGEMENTS

Thanks to chromatic and Michael G Schwern for the excellent Test::Builder,
without which this module wouldn't be possible.

Thanks to Adrian Howard for writing Test::Exception, from which most of
this module is taken.

=head1 COPYRIGHT & LICENSE

Copyright 2005-2018 Andy Lester.

This program is free software; you can redistribute it and/or modify
it under the terms of the Artistic License v2.0.

http://www.opensource.org/licenses/Artistic-2.0

Please note that these modules are not products of or supported by the
employers of the various contributors to the code.

=head1 AUTHOR

Andy Lester, C<andy@petdance.com>

=cut

1;


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker qw( WriteMakefile );
use 5.006001;

if ( not eval { require LWP::Simple; 1; } ) {
    print <<'EOF';

NOTE: It seems that you don't have LWP::Simple installed.
      The weblint program will not be able to retrieve web pages.

EOF
}

my %parms = (
    NAME            => 'HTML::Lint',
    DISTNAME        => 'HTML-Lint',
    VERSION_FROM    => 'lib/HTML/Lint.pm',
    ABSTRACT_FROM   => 'lib/HTML/Lint.pm',
    PMLIBDIRS       => [qw(lib/)],
    AUTHOR          => 'Andy Lester <andy@petdance.com>',
    MIN_PERL_VERSION=> 5.006,
    PREREQ_PM       => {
        'Exporter'              => 0,
        'Test::More'            => 0,
        'Test::Builder'         => 0,
        'Test::Builder::Tester' => 0,

        'File::Find'            => 0,
        'HTML::Entities'        => 0,
        'HTML::Parser'          => '3.47',
        'HTML::Tagset'          => '3.03',
    },
    EXE_FILES   => [qw(bin/weblint)],
    dist => {
        COMPRESS => 'gzip -9f',
        SUFFIX => 'gz',
    },
    clean => { FILES => 'HTML-Lint-*' },
);

if ( $ExtUtils::MakeMaker::VERSION =~ /^\d[.]\d\d$/ and $ExtUtils::MakeMaker::VERSION > 6.30 ) {
    $parms{LICENSE} = 'artistic_2';
}

if ( $ExtUtils::MakeMaker::VERSION ge '6.46' ) {
    $parms{META_ADD} = {
        resources => {
            homepage   => 'http://search.cpan.org/dist/html-lint',
            bugtracker => 'https://github.com/petdance/html-lint/issues',
            license    => 'http://www.opensource.org/licenses/artistic-license-2.0.php',
            repository => 'https://github.com/petdance/html-lint',
        },
    };
}

WriteMakefile( %parms );

sub MY::postamble { ## no critic ( Subroutines::ProhibitQualifiedSubDeclarations )
    my $postamble = <<'MAKE_FRAG';

.PHONY: tags critic

tags:
	ctags -f tags --recurse --totals \
		--exclude=blib \
		--exclude=.svn \
		--exclude='*~' \
		--languages=Perl --langmap=Perl:+.t \

critic:
	perlcritic -1 -q -profile perlcriticrc lib/ bin/weblint Makefile.PL t/*.t

PROF_ARGS = -Mblib blib/script/weblint index.html

timed: all
	$(PERL) $(PROF_ARGS) >> /dev/null 2>&1

dprof: all
	$(PERL) -d:DProf $(PROF_ARGS) >> /dev/null 2>&1
	dprofpp -R

dproflb: all
	$(PERL) -d:DProfLB $(PROF_ARGS) >> /dev/null 2>&1
	dprofpp -R

fastprof: all
	$(PERL) -d:FastProf $(PROF_ARGS) >> /dev/null 2>&1
	fprofpp

profile: all
	$(PERL) -d:Profile $(PROF_ARGS) >> /dev/null 2>&1
	less prof.out

profiler: all
	$(PERL) -MDevel::Profiler $(PROF_ARGS) >> /dev/null 2>&1
	dprofpp -R

smallprof: all
	$(PERL) -d:SmallProf $(PROF_ARGS) >> /dev/null 2>&1
	sort -k 2nr,2 smallprof.out | less

nytprof: all
	$(PERL) -d:NYTProf $(PROF_ARGS) >> /dev/null 2>&1
	nytprofhtml

MAKE_FRAG

    return $postamble;
}


lib/HTML/Lint/Parser.pm
package HTML::Lint::Parser;

use warnings;
use strict;

use HTML::Parser 3.20;
use HTML::Tagset 3.03;

use HTML::Lint::Error ();
use HTML::Lint::HTML4 qw( %isKnownAttribute %isRequired %isNonrepeatable %isObsolete );
use HTML::Entities qw( %char2entity %entity2char );

use parent 'HTML::Parser';

=head1 NAME

HTML::Lint::Parser - Parser for HTML::Lint.  No user-serviceable parts inside.

=head1 VERSION

Version 2.28

=cut

our $VERSION = '2.28';

=head1 SYNOPSIS

See L<HTML::Lint> for all the gory details.

=head1 METHODS

=head2 new( $gripe )

Constructor for the main parsing object.  The I<$gripe> argument
is a coderef to a function that can handle errors from the parser.
It is only ever (so far) C<HTML::Lint::gripe()>.

=cut

sub new {
    my $class = shift;
    my $gripe = shift;

    my $self =
        HTML::Parser->new(
            api_version => 3,
            start_document_h   => [ \&_start_document, 'self' ],
            end_document_h     => [ \&_end_document,   'self,line,column' ],
            start_h            => [ \&_start,          'self,tagname,line,column,@attr' ],
            end_h              => [ \&_end,            'self,tagname,line,column,tokenpos,@attr' ],
            comment_h          => [ \&_comment,        'self,tagname,line,column,text' ],
            text_h             => [ \&_text,           'self,text' ],
            strict_names       => 0,
            empty_element_tags => 1,
            attr_encoded       => 1,
        );
    bless $self, $class;

    $self->{_gripe}      = $gripe;
    $self->{_stack}      = [];
    $self->{_directives} = {};

    return $self;
}

=head2 $parser->gripe( $errorcode, [ arg1=>val1, ...] )

Calls the passed-in gripe function.

If a given directive has been set to turn off a given message, then
the parent gripe never gets called.

=cut

sub gripe {
    my $self      = shift;
    my $errorcode = shift;

    if ( $self->_displayable( $errorcode ) ) {
        $self->{_gripe}->( $errorcode, @_ );
    }

    return;
}

sub _displayable {
    my $self      = shift;
    my $errorcode = shift;

    my $directives = $self->{_directives};
    if ( not defined $directives->{$errorcode} ) {
        return 1;
    }
    else {
        return $directives->{$errorcode};
    }
}

sub _start_document {
    return;
}

sub _end_document {
    my ($self,$line,$column) = @_;

    for my $tag ( sort keys %isRequired ) {
        if ( !$self->{_first_seen}->{$tag} ) {
            $self->gripe( 'doc-tag-required', tag => $tag );
        }
    }

    return;
}

sub _start {
    my ($self,$tag,$line,$column,@attr) = @_;

    $self->{_line} = $line;
    $self->{_column} = $column;

    my $validattr = $isKnownAttribute{ $tag };
    if ( $validattr ) {
        my %seen;
        my $i = 0;
        while ( $i < @attr ) {
            my ($attr,$val) = @attr[$i++,$i++];
            if ( $seen{$attr}++ ) {
                $self->gripe( 'attr-repeated', tag => $tag, attr => $attr );
            }

            if ( !$validattr->{$attr} ) {
                $self->gripe( 'attr-unknown', tag => $tag, attr => $attr );
            }

            $self->_entity($val, 'attr');
        } # while attribs
    }
    else {
        $self->gripe( 'elem-unknown', tag => $tag );
    }
    $self->_element_push( $tag ) unless $HTML::Tagset::emptyElement{ $tag };

    if ( my $where = $self->{_first_seen}{$tag} ) {
        if ( $isNonrepeatable{$tag} ) {
            $self->gripe( 'elem-nonrepeatable',
                            tag => $tag,
                            where => HTML::Lint::Error::where( @{$where} )
                        );
        }
    }
    else {
        $self->{_first_seen}{$tag} = [$line,$column];
    }

    # Call any other overloaded func
    my $tagfunc = "_start_$tag";
    if ( $self->can($tagfunc) ) {
        $self->$tagfunc( $tag, @attr );
    }

    return;
}

sub _text {
    my ($self,$text) = @_;

    $self->_entity($text, 'text');

    return;
}

sub _entity {
    my ($self,$text,$type) = @_;

    if ( not $self->{_entity_lookup} ) {
        my @entities = sort keys %HTML::Entities::entity2char;
        # Strip his semicolons
        s/;$// for @entities;
        $self->{_entity_lookup} = { map { ($_,1) } @entities };
    }

    while ( $text =~ /([^\x09\x0A\x0D -~])/g ) {
        my $bad = $1;
        $self->gripe(
            $type . '-use-entity',
                char => sprintf( '\x%02lX', ord($bad) ),
                entity => $char2entity{ $bad } || '&#' . ord($bad) . ';',
        );
    }

    while ( $text =~ /&([^ ;]*;?)/g ) {
        my $match = $1;

        if ( $match eq '' ) {
            $self->gripe( $type . '-use-entity', char => '&', entity => '&amp;' );
        }
        elsif ( $match !~ m/;$/ ) {
            if ( exists $self->{_entity_lookup}->{$match}
                 || $match =~ m/^#(\d+)$/ || $match =~ m/^#x[\dA-F]+$/i) {
                $self->gripe( $type . '-unclosed-entity', entity => "&$match;" );
            }
            else {
                $self->gripe( $type . '-unknown-entity', entity => "&$match" );
            }
        }
        elsif ( $match =~ m/^#(\d+);$/ ) {
            if ( $1 > 65536 ) {
                $self->gripe( $type . '-invalid-entity', entity => "&$match" );
            }
        }
        elsif ( $match =~ m/^#x([\dA-F]+);$/i ) {
            if ( length($1) > 4 ) {
                $self->gripe( $type . '-invalid-entity', entity => "&$match" );
            }
        }
        else {
            $match =~ s/;$//;
            if ( !exists $self->{_entity_lookup}->{$match} ) {
                $self->gripe( $type . '-unknown-entity', entity => "&$match;" );
            }
        }
    }

    return;
}

sub _comment {
    my ($self,$tagname,$line,$column,$text) = @_;

    # Look for the html-lint directives
    if ( $tagname =~ m/^\s*html-lint\s*(.+)\s*$/ ) {
        my $text = $1;

        my @commands = split( /\s*,\s*/, $text );

        for my $command ( @commands ) {
            my ($directive,$value) = split( /\s*:\s*/, $command, 2 );
            _trim($_) for ($directive,$value);

            if ( ($directive ne 'all') &&
                ( not exists $HTML::Lint::Error::errors{ $directive } ) ) {
                $self->gripe( 'config-unknown-directive',
                    directive => $directive,
                    where     => HTML::Lint::Error::where($line,$column)
                );
                next;
            }

            my $normalized_value = _normalize_value( $value );
            if ( !defined($normalized_value) ) {
                $self->gripe( 'config-unknown-value',
                    directive => $directive,
                    value     => $value,
                    where     => HTML::Lint::Error::where($line,$column)
                );
                next;
            }

            if ( $directive eq 'all' ) {
                for my $err ( keys %HTML::Lint::Error::errors ) {
                    $self->_set_directive( $err, $normalized_value );
                }
            }
            else {
                $self->_set_directive( $directive, $normalized_value );
            }
        }
    }

    return;
}

sub _set_directive {
    my $self  = shift;
    my $which = shift;
    my $what  = shift;

    $self->{_directives}{$which} = $what;

    return;
}


sub _normalize_value {
    my $what = shift;

    $what = _trim( $what );
    return 1 if $what eq '1' || $what eq 'on'  || $what eq 'true';
    return 0 if $what eq '0' || $what eq 'off' || $what eq 'false';
    return undef;
}

sub _trim {
    $_[0] =~ s/^\s+//;
    $_[0] =~ s/\s+$//;

    return $_[0];
}

sub _end {  ## no critic ( Subroutines::ProhibitManyArgs ) I have no choice in what these args are.
    my ($self,$tag,$line,$column,$tokenpos,@attr) = @_;

    $self->{_line} = $line;
    $self->{_column} = $column;

    if ( !$tokenpos ) {
        # This is a dummy end event for something like <img />.
        # Do nothing.
    }
    elsif ( $HTML::Tagset::emptyElement{ $tag } ) {
        $self->gripe( 'elem-empty-but-closed', tag => $tag );
    }
    else {
        if ( $self->_in_context($tag) ) {
            my @leftovers = $self->_element_pop_back_to($tag);
            for ( @leftovers ) {
                my ($tag,$line,$col) = @{$_};
                $self->gripe( 'elem-unclosed', tag => $tag,
                        where => HTML::Lint::Error::where($line,$col) )
                        unless $HTML::Tagset::optionalEndTag{$tag};
            } # for
        }
        else {
            $self->gripe( 'elem-unopened', tag => $tag );
        }
    } # is empty element

    # Call any other overloaded func
    my $tagfunc = "_end_$tag";
    if ( $self->can($tagfunc) ) {
        $self->$tagfunc( $tag, $line );
    }

    return;
}

sub _element_push {
    my $self = shift;
    for ( @_ ) {
        push( @{$self->{_stack}}, [$_,$self->{_line},$self->{_column}] );
    } # while

    return;
}

sub _find_tag_in_stack {
    my $self = shift;
    my $tag = shift;
    my $stack = $self->{_stack};

    my $offset = @{$stack} - 1;
    while ( $offset >= 0 ) {
        if ( $stack->[$offset][0] eq $tag ) {
            return $offset;
        }
        --$offset;
    } # while

    return;
}

sub _element_pop_back_to {
    my $self = shift;
    my $tag = shift;

    my $offset = $self->_find_tag_in_stack($tag) or return;

    my @leftovers = splice( @{$self->{_stack}}, $offset + 1 );
    pop @{$self->{_stack}};

    return @leftovers;
}

sub _in_context {
    my $self = shift;
    my $tag = shift;

    my $offset = $self->_find_tag_in_stack($tag);
    return defined $offset;
}

# Overridden tag-specific stuff
sub _start_img {    ## no critic ( Subroutines::ProhibitUnusedPrivateSubroutines )  # Called by parser based on tag name.
    my ($self,$tag,%attr) = @_;

    my ($h,$w,$src) = @attr{qw( height width src )};
    if ( defined $h && defined $w ) {
        # Check sizes
    }
    else {
        $self->gripe( 'elem-img-sizes-missing', src=>$src );
    }
    if ( not defined $attr{alt} ) {
        $self->gripe( 'elem-img-alt-missing', src=>$src );
    }

    return;
}

sub _start_input {  ## no critic ( Subroutines::ProhibitUnusedPrivateSubroutines )  # Called by parser based on tag name.
    my ($self,$tag,%attr) = @_;

    my ($type,$alt) = @attr{qw( type alt )};
    if ( defined($type) && (lc($type) eq 'image') ) {
        my $ok = defined($alt);
        if ( $ok ) {
            $alt =~ s/^ +//;
            $alt =~ s/ +$//;
            $ok = ($alt ne '');
        }
        if ( !$ok ) {
            my $name = $attr{name};
            $name = '' unless defined $name;
            $self->gripe( 'elem-input-alt-missing', name => $name );
        }
    }

    return;
}

1;
